File: lib/dns_stack.py
"""Route 53 DNS with health checks and failover."""

from constructs import Construct
from cdktf_cdktf_provider_aws.route53_zone import Route53Zone
from cdktf_cdktf_provider_aws.route53_record import Route53Record
from cdktf_cdktf_provider_aws.route53_health_check import Route53HealthCheck


class DnsStack(Construct):
    """Route 53 DNS with health checks and failover routing."""

    def __init__(self, scope: Construct, construct_id: str, environment_suffix: str,
                 primary_provider, primary_lambda_url: str, secondary_lambda_url: str):
        super().__init__(scope, construct_id)

        # Hosted zone
        hosted_zone = Route53Zone(
            self, "hosted_zone",
            name=f"payment-{environment_suffix}.testing.local",
            comment=f"Payment processing system - {environment_suffix}",
            tags={"Name": f"payment-zone-{environment_suffix}"},
            provider=primary_provider,
        )

        # Clean Lambda URLs
        primary_url = primary_lambda_url.replace("https://", "").rstrip("/")
        secondary_url = secondary_lambda_url.replace("https://", "").rstrip("/")

        # Health checks
        primary_health = Route53HealthCheck(
            self, "primary_health",
            type="HTTPS",
            resource_path="/",
            fqdn=primary_url,
            port=443,
            request_interval=30,
            failure_threshold=3,
            measure_latency=True,
            tags={"Name": f"payment-primary-health-{environment_suffix}"},
            provider=primary_provider,
        )

        secondary_health = Route53HealthCheck(
            self, "secondary_health",
            type="HTTPS",
            resource_path="/",
            fqdn=secondary_url,
            port=443,
            request_interval=30,
            failure_threshold=3,
            measure_latency=True,
            tags={"Name": f"payment-secondary-health-{environment_suffix}"},
            provider=primary_provider,
        )

        # Failover records
        Route53Record(
            self, "primary_record",
            zone_id=hosted_zone.zone_id,
            name=f"api.payment-{environment_suffix}.testing.local",
            type="CNAME",
            ttl=60,
            records=[primary_url],
            set_identifier="primary",
            health_check_id=primary_health.id,
            failover_routing_policy={"type": "PRIMARY"},
            provider=primary_provider,
        )

        Route53Record(
            self, "secondary_record",
            zone_id=hosted_zone.zone_id,
            name=f"api.payment-{environment_suffix}.testing.local",
            type="CNAME",
            ttl=60,
            records=[secondary_url],
            set_identifier="secondary",
            health_check_id=secondary_health.id,
            failover_routing_policy={"type": "SECONDARY"},
            provider=primary_provider,
        )

        self.hosted_zone_id = hosted_zone.zone_id

File: archive/pulumi-ts/Pr6318/lib/dns-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface DnsStackArgs {
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
  primaryAlbDnsName: pulumi.Output<string>;
  drAlbDnsName: pulumi.Output<string>;
  primaryAlbZoneId: pulumi.Output<string>;
  drAlbZoneId: pulumi.Output<string>;
  primaryProvider: aws.Provider;
}

export class DnsStack extends pulumi.ComponentResource {
  public readonly primaryEndpoint: pulumi.Output<string>;
  public readonly drEndpoint: pulumi.Output<string>;
  public readonly hostedZoneId: pulumi.Output<string>;

  constructor(
    name: string,
    args: DnsStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:dns:DnsStack', name, args, opts);

    const {
      environmentSuffix,
      tags,
      primaryAlbDnsName,
      drAlbDnsName,
      primaryAlbZoneId,
      drAlbZoneId,
      primaryProvider,
    } = args;

    // Create hosted zone in primary region
    const hostedZone = new aws.route53.Zone(
      `payments-zone-${environmentSuffix}`,
      {
        name: `payments-${environmentSuffix}.internal`,
        comment: 'Managed by Pulumi for DR setup',
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `payments-zone-${environmentSuffix}`,
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Health check for primary ALB
    const primaryHealthCheck = new aws.route53.HealthCheck(
      `primary-hc-${environmentSuffix}`,
      {
        type: 'HTTP',
        resourcePath: '/health.html',
        fqdn: primaryAlbDnsName,
        port: 80,
        requestInterval: 30,
        failureThreshold: 3,
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `primary-hc-${environmentSuffix}`,
          'DR-Role': 'primary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Health check for DR ALB
    const drHealthCheck = new aws.route53.HealthCheck(
      `dr-hc-${environmentSuffix}`,
      {
        type: 'HTTP',
        resourcePath: '/health.html',
        fqdn: drAlbDnsName,
        port: 80,
        requestInterval: 30,
        failureThreshold: 3,
        tags: pulumi.all([tags]).apply(([t]) => ({
          ...t,
          Name: `dr-hc-${environmentSuffix}`,
          'DR-Role': 'secondary',
        })),
      },
      { provider: primaryProvider, parent: this }
    );

    // Primary failover record
    new aws.route53.Record(
      `primary-record-${environmentSuffix}`,
      {
        zoneId: hostedZone.zoneId,
        name: `payments-${environmentSuffix}.internal`,
        type: 'A',
        setIdentifier: `primary-${environmentSuffix}`,
        failoverRoutingPolicies: [
          {
            type: 'PRIMARY',
          },
        ],
        aliases: [
          {
            name: primaryAlbDnsName,
            zoneId: primaryAlbZoneId,
            evaluateTargetHealth: true,
          },
        ],
        healthCheckId: primaryHealthCheck.id,
      },
      { provider: primaryProvider, parent: this }
    );

    // DR failover record
    new aws.route53.Record(
      `dr-record-${environmentSuffix}`,
      {
        zoneId: hostedZone.zoneId,
        name: `payments-${environmentSuffix}.internal`,
        type: 'A',
        setIdentifier: `dr-${environmentSuffix}`,
        failoverRoutingPolicies: [
          {
            type: 'SECONDARY',
          },
        ],
        aliases: [
          {
            name: drAlbDnsName,
            zoneId: drAlbZoneId,
            evaluateTargetHealth: true,
          },
        ],
        healthCheckId: drHealthCheck.id,
      },
      { provider: primaryProvider, parent: this }
    );

    this.primaryEndpoint = pulumi.interpolate`http://${primaryAlbDnsName}`;
    this.drEndpoint = pulumi.interpolate`http://${drAlbDnsName}`;
    this.hostedZoneId = hostedZone.zoneId;

    this.registerOutputs({
      primaryEndpoint: this.primaryEndpoint,
      drEndpoint: this.drEndpoint,
      hostedZoneId: this.hostedZoneId,
    });
  }
}
